Rework Introduction of the Installation guide

gdemonet opened the pull request:
**Component**: docs

**Context**:

We need to provide clearer information to users when they start reading the installation guide, to better understand the various architectures they can deploy.

**Summary**:

- Replace the previous "Installation Plan" section with a few example architectures and recommendations on how to define one
- References to the previous "quickstart" architecture were removed
- Add sizing notes in the introduction (obtained from monitoring data from an OpenStack deployment)
- Improve roles and taints explanation

**Acceptance criteria**:

- Wording is correct and explicit
- Information is valid and accurate
- Doc builds and displays properly

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2112 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `documentation/2112-revamp-install-plan` into
`development/2.4`
* [w/2.5/documentation/2112-revamp-install-plan](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/documentation/2112-revamp-install-plan) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

bert-e commented:
# History mismatch



Merge commit #54e27d43b6dce91132bbecaad08fbdb7f5b5f474 on the integration branch
`w/2.5/documentation/2112-revamp-install-plan` is merging a branch which is neither the current
branch `documentation/2112-revamp-install-plan` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `documentation/2112-revamp-install-plan` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard commented on docs/installation/introduction.rst:
```diff
@@ -1,6 +1,104 @@
+Choosing a Deployment Architecture
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Before starting the installation, choosing an architecture is recommended, as
+it can impact sizing of your machines and other infrastructure-related details.
```
`s/your/the/`

alexandre-allard commented on docs/installation/setup.rst:
```diff
@@ -127,14 +125,17 @@ For more detail(s), refer to the official Red Hat documentation:
 :term:`etcd`
 ^^^^^^^^^^^^
-For production environments, a dedicated block device for :term:`etcd` is
+For production environments, a dedicated block device for ``etcd` is
```
A back tick is missing after `etcd`

gdemonet committed:
docs/install: Rework reference architectures

The single architecture that used to be in the Quickstart guide does not
provide sufficient information to readers about the flexibility
available when choosing an architecture for their clusters.
We replace the previous "Installation Plan" section with a few example
architecture and recommendations on how to define one.

References to the previous "quickstart" architecture were removed.

Issue: #2112

gdemonet committed:
docs/install: Fix few casing issues

gdemonet committed:
docs/install: Add sizing notes in the introduction

Using monitoring data from an OpenStack deployment, we evaluate the
resource requirements of the services included in a standard MetalK8s
installation. These numbers are slightly over the actually observed
consumption, but shouldn't be taken as too conservative either.

Issue: #2112

gdemonet committed:
docs/install: Link to `etcd` role

The setup of /var/lib/etcd partition didn't mention that it's only
required on Nodes with the `etcd` role.
We also fix a few formatting issues (notably, do not link every
occurence of `etcd` to the Glossary).

Issue: #2112

gdemonet committed:
docs/install: Improve roles and taints explanation

We link to the few example architectures to better explain how users
should set roles and taints to define their clusters.
We also remove mention of the "node" role, which was confusing and not
used anyway.

Issue: #2112

Ebaneck commented on docs/installation/expansion.rst:
```diff
@@ -19,20 +19,14 @@ from a single entrypoint. This operation can be done either through
+architecture.
+
+List the machines to deploy and their associated roles, and deploy each of them
+using the following process, either from the GUI or CLI. Note however, that
```
Could be nice to make `GUI or CLI` HTTP links such that users can click to follow it on another tab

Ebaneck commented on docs/installation/introduction.rst:
```diff
@@ -1,6 +1,104 @@
+
+Standard Architecture
+"""""""""""""""""""""
+MetalK8s has been designed with on-premise, offline, reliable deployments in
```
Since we should be talking about the present(now) and presumably looking into the future, knowing that we will continue to follow these design pattern; this could read...

`MetalK8s is designed with on-premise, offline ....`

No??

thomasdanan commented on docs/installation/introduction.rst:
```diff
@@ -1,6 +1,104 @@
+the following steps:
+
+#. :doc:`Setup <./setup>` of the environment (with requirements and example
+   OpenStack deployment)
```
Not sure we want to talk about Openstack here? since it is not where MetalK8s will be installed on-prem, in prod.

thomasdanan commented on docs/installation/introduction.rst:
```diff
@@ -1,6 +1,104 @@
+#. :doc:`Deployment <./bootstrap>` of the :term:`Bootstrap node`, the first
+   machine in the cluster
+#. :doc:`Expansion <./expansion>` of the cluster, orchestrated from the
+   Bootstrap node
```
And I would add Post Installation, especially to provision prom / alert manager volumes in order to have a fully functional MetalK8s cluster

thomasdanan commented on docs/installation/introduction.rst:
```diff
@@ -1,6 +1,104 @@
+The first example architecture proposed follows this spirit, focusing on
+reliability rather than compacity:
+
+- One machine dedicated to running Bootstrap services (see
```
When deploying the bootstrap node, it automatically deploys services related to master or etcd so achieving this architecture would mean to remove some roles/taints, it is really the standard arch we want to recommend? shouldn't we suggest a standard arch where bootstrap node is also having matser/etcd role?

gdemonet replied:
I think we prefer running bootstrap services independently, to prevent Salt Master from having an impact on control plane services. The fact that Bootstrap runs everything after the initial installation is just a constraint (you need it to start expanding the cluster). So I don't know, maybe I can reword the "standard" into something like "extended" / "flat". WDYT?

thomasdanan replied:
I would like to have @xaf-scality opinion on this one ...

gdemonet replied:
Had a face to face conversation with him: we'll use the 3-node control-plane + infra, including bootstrap, as the "standard", and suggest the "extended" as a third option.

thomasdanan commented on docs/installation/introduction.rst:
```diff
@@ -1,6 +1,104 @@
+  100+ nodes) for running the Kubernetes_ control plane (with
+  :ref:`core K8s services<node-role-master>` and the backing
+  :ref:`etcd DB<node-role-etcd>`)
+- One or more machines dedicated to running Infra services (see
```
Shouldn't we recommend to have at least 2 or 3 nodes for infra services? to have a minimum of resilience in case of HW failure?

gdemonet replied:
We can, but for now, this requires manual changes that won't be persisted on upgrade.

thomasdanan replied:
Okay understood

thomasdanan commented on docs/installation/introduction.rst:
```diff
@@ -1,6 +1,104 @@
+node can be configured to also allow running applications next to all the other
+services required (see :ref:`the section about taints<node-taints>` below).
+
+A single node cluster does not provide any form of resilience to machine or
```
I realize we are not saying that MetalK8s deployment is not designed for multi site architecture (at least for now). We might want to highlight it at the beginning of this chapter.

gdemonet replied:
OK I'll add a note about this.

thomasdanan commented on docs/installation/introduction.rst:
```diff
@@ -1,6 +1,104 @@
+expected load, and the exact impact of colocating an application with MetalK8s
+services needs to be evaluated by said application's provider.
+
+Variations
```
I would think that this architecture is the standard/default one we should recommend. It is also the easiest one to achieve using the tooling and UI we have today. @xaf-scality shouldn't we suggest this one as the default one?

thomasdanan commented on docs/installation/introduction.rst:
```diff
@@ -14,7 +112,7 @@ Nodes
 containers and can be managed by the cluster (control plane services,
```
I realized Node may be confusing when you think about Nodes having Node role no?

gdemonet replied:
The `node-role.kubernetes.io/node` role label doesn't really do much, and indeed, is slightly confusing. But we can't change the core K8s names at our will, so I don't see what kind of improvement could be added here.

thomasdanan replied:
Yes I am not sure how to improve as well ...

gdemonet replied:
I think we can omit it entirely. I haven't heard of workloads using this `nodeSelector` anyway, as most wouldn't use one at all, and be scheduled anywhere not tainted. I'll adjust the schemas to represent "workload services" instead of the confusing "node" role.

thomasdanan commented on docs/installation/introduction.rst:
```diff
@@ -80,6 +188,10 @@ MetalK8s uses five different **roles**, that may be combined freely:
+
 .. _node-taints:
 
 Node Taints
```
By reading at the node-taints chapter, I am not sure to understand what taints I can set and how to set it ...

gdemonet replied:
OK, I'll expand a bit with an example or two.

gdemonet replied:
I added an example and a link to the official doc. Is that enough?

thomasdanan commented on docs/installation/introduction.rst:
```diff
@@ -119,32 +247,59 @@ other ranges during the
+  and container registries, need a minimum of 1 CPU core and 2 GB of RAM
+- Control plane services, including Kubernetes API and its backing ``etcd``
+  database, require at least 0.5 CPU core and 1.5 GB of RAM
+- Infra services, running Prometheus and Alertmanager among others, need 0.5
```
Shouldn't we add some sizing for any node in the cluster? even if no bootstrap, control plane or infra services is running on it?

thomasdanan replied:
maybe a little table where you see type of services, associated role and corresponding sizing (cpu, ram and storage) would be even more clear

gdemonet replied:
> Shouldn't we add some sizing for any node in the cluster?

The whole point is that we **cannot** know what sizing is required to run "any Node". That's why we explicitly state that users should _Refer to the documentation of the applications planned to run in the deployed cluster before completing the sizing, as their needs will compete with the cluster's_

> maybe a little table [...] would be even more clear

Will do.

thomasdanan replied:
Obviously the sizing will depend on what application is running on it but the sizing due to MetalK8s services running on it (i.e. node exporter, kubelet, salt minion I guess) is either negligible either we could / should provide an estimation for it so that it is taken in account when building the global sizing of the cluster.

gdemonet replied:
I think they are negligible compared to the actual workloads you'll want to deploy on top of MetalK8s, but I can try to provide some sizing info.

@xaf-scality also suggested giving capacity required for the etcd + prometheus + alertmanager partitions, and add a recommendation for using a dedicated /var partition of 100+ GB

thomasdanan commented on docs/installation/introduction.rst:
```diff
@@ -119,32 +247,59 @@ other ranges during the
+Installing MetalK8s on virtual machines in cloud environments is perfectly
+achievable. Note however that most cloud providers have their own offerings
+for hosted Kubernetes clusters, which can save time and efforts (operation
+of the cluster being delegated).
```
I would remove this sentence as the one reading this doc wants to install MetalK8s and probably made his choice between hosted K8s services and MetalK8s

gdemonet replied:
OK, will remove.

thomasdanan commented on docs/installation/post-install.rst:
```diff
@@ -1,9 +1,9 @@
 Post-Installation Procedure
```
Not related to your changes, but I would remove sparse file example as we don't want to use it in prod ... and replace it with real device example.

gdemonet replied:
Will do when tackling #2117

thomasdanan reviewed (commented):
Great doc !!!

gdemonet committed:
docs/install: Add post install to the plan summary

Issue: #2112

gdemonet committed:
docs/install: Add note to remove Bootstrap taints

Issue: #2112

gdemonet committed:
docs/install: Add warning about multi-sites

Since we don't want to recommend multi-sites clusters managed by a
single K8s control plane (as it has been observed to perform poorly), we
warn users about it.

Issue: #2112

gdemonet committed:
docs/install: Add note about cloud deployments

Issue: #2112

lucieleonard commented on docs/installation/introduction.rst:
```diff
@@ -1,6 +1,125 @@
 
+Foreword
+^^^^^^^^
+MetalK8s is a Kubernetes_ distribution with a number of addons carefully
```
add-ons

lucieleonard replied:
"carefully picked" == chosen?
Might be removed.

gdemonet replied:
I'll rephrase with `add-ons selected for on-premises deployments, ...`

lucieleonard commented on docs/installation/introduction.rst:
```diff
@@ -1,6 +1,125 @@
+
+Standard Architecture
+"""""""""""""""""""""
+The recommended architecture when installing a small size MetalK8s cluster
```
small-sized MetalK8s cluster

lucieleonard commented on docs/installation/introduction.rst:
```diff
@@ -1,6 +1,125 @@
+
+Please note that sizing of such compact clusters needs to account for the
+expected load, and the exact impact of colocating an application with MetalK8s
+services needs to be evaluated by said application's provider.
```
can be transformed into a .. note:: 
"Sizing of such compact clusters..."

lucieleonard commented on docs/installation/introduction.rst:
```diff
@@ -1,6 +1,125 @@
+Before starting the installation, choosing an architecture is recommended, as
+it can impact sizing of the machines and other infrastructure-related details.
+
+.. note:: "Machines" may indicate bare-metal servers or VMs interchangeably
```
. at the end

lucieleonard commented on docs/installation/introduction.rst:
```diff
@@ -1,6 +1,125 @@
+
+Machines dedicated to the control plane do not need large
+amounts of resources (see
+:ref:`the sizing notes below<installation-intro-sizing>`, and can safely run as
```
close ')'

lucieleonard commented on docs/installation/introduction.rst:
```diff
@@ -1,6 +1,125 @@
+- One machine dedicated to running Bootstrap services (see
+  :ref:`the Bootstrap role<node-role-bootstrap>` definition below)
+- Three extra machines (or five if installing a really large cluster,
+  100+ nodes) for running the Kubernetes_ control plane (with
```
large cluster, e.g. >100 nodes)

lucieleonard commented on docs/installation/introduction.rst:
```diff
@@ -1,6 +1,125 @@
+  :ref:`etcd DB<node-role-etcd>`)
+- One or more machines dedicated to running Infra services (see
+  :ref:`the Infra role<node-role-infra>`)
+- Any number of machines dedicated to running applications, the number and
```
Any == too vague. Can we give a range? Or suggest a number?

gdemonet replied:
No we can't really, it depends on what applications you'll want to run...

lucieleonard commented on docs/installation/introduction.rst:
```diff
@@ -1,6 +1,125 @@
+
+Extended Architecture
+"""""""""""""""""""""
+MetalK8s is designed with on-premise, offline, reliable deployments in mind.
```
Does this apply to MetalK8s in general? If so, why stating it here? (if it does apply only for this architecture, leave this sentence. Else, I suggest removing it/moving it to the introduction.)

gdemonet replied:
Indeed, no reason to state it here again.

lucieleonard commented on docs/installation/introduction.rst:
```diff
@@ -1,6 +1,125 @@
+""""""""""
+It is possible to customize the chosen architecture using combinations of
+:ref:`roles<node-roles>` and :ref:`taints<node-taints>`, which are described
+below, to adapt to the available infrastructure.
```
Customize the chosen architecture to adapt the available infrastructure using combinations of :ref:`roles<node-roles>` and :ref:`taints<node-taints>`.
Described below == reference/link? Does it refers to node taints and roles or the two paragraphs below?

gdemonet replied:
The references are set on `roles` and `taints` already. Not clear enough?

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

gdemonet commented:
/approve

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `documentation/2112-revamp-install-plan` into
`development/2.4`
* [w/2.5/documentation/2112-revamp-install-plan](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/documentation/2112-revamp-install-plan) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @ebaneck

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @ebaneck







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort










*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #aefcb26f857757791c02cbb9323e57ab6542e262 on the integration branch
`w/2.5/documentation/2112-revamp-install-plan` is merging a branch which is neither the current
branch `documentation/2112-revamp-install-plan` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `documentation/2112-revamp-install-plan` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `documentation/2112-revamp-install-plan` into
`development/2.4`
* [w/2.5/documentation/2112-revamp-install-plan](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/documentation/2112-revamp-install-plan) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort










*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/12eca57a5868bf61320c1b4151800ac4e22d41ea) did not succeed in branch w/2.5/documentation/2112-revamp-install-plan.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye gdemonet.
